[inductor] use smaller RBLOCK for expensive reduction kernels

shunting314 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* #126348
* __->__ #126477

Triton sometimes uses less registers for more expensive kernel which results in worse perf ( https://github.com/pytorch/pytorch/issues/126463 ). This may make inductor end up with a sub-optimal config. Use a smaller max RBLOCK if the reduction potentially need many registers.

Will run perf test..


cc @voznesenskym @penguinwu @EikanWang @jgong5 @Guobing-Chen @XiaobingSuper @zhuhaozhe @blzheng @wenzhe-nrv @jiayisunx @peterbell10 @ipiszy @yf225 @chenyang78 @kadeng @muchulee8 @ColinPeppler @amjames @desertfire @chauhang

shunting314 committed:
[inductor] use smaller RBLOCK for expensive reduction kernels

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] use smaller RBLOCK for expensive reduction kernels"




cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 committed:
Update on "[inductor] use smaller RBLOCK for expensive reduction kernels"


Triton sometimes uses less registers for more expensive kernel which results in worse perf ( https://github.com/pytorch/pytorch/issues/126463 ). This may make inductor end up with a sub-optimal config. Use a smaller max RBLOCK if the reduction potentially need many registers.

Will run perf test..


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

eellison reviewed (commented):
Maybe this motivation to do live variable analysis.. going to wait to see how perf results look.

shunting314 commented on torch/_inductor/runtime/triton_heuristics.py:
```diff
@@ -1480,12 +1485,32 @@ def _reduction_configs(
     rnumel = size_hints[-1]
 
+    MAX_RBLOCK = 2048
+    if (
```
no. I checked `CompiledKernel.n_spills`, it's 0.

shunting314 committed:
Update on "[inductor] use smaller RBLOCK for expensive reduction kernels"


Triton sometimes uses less registers for more expensive kernel which results in worse perf ( https://github.com/pytorch/pytorch/issues/126463 ). This may make inductor end up with a sub-optimal config. Use a smaller max RBLOCK if the reduction potentially need many registers.

Will run perf test..


cc voznesenskym penguinwu EikanWang jgong5 Guobing-Chen XiaobingSuper zhuhaozhe blzheng wenzhe-nrv jiayisunx peterbell10 ipiszy yf225 chenyang78 kadeng muchulee8 ColinPeppler amjames desertfire chauhang

[ghstack-poisoned]

shunting314 commented:
Perf test on this PR is [roughly neutral](https://hud.pytorch.org/benchmark/compilers?startTime=Wed%2C%2015%20May%202024%2018%3A49%3A51%20GMT&stopTime=Wed%2C%2022%20May%202024%2018%3A49%3A51%20GMT&granularity=hour&suite=torchbench&mode=training&dtype=amp&lBranch=gh/shunting314/146/head&lCommit=32e418d539d0a70f95ef4acb2abec0013fc160da&rBranch=main&rCommit=53f73cdeb6fe53155f91064ca15b722e80dec2f3). 

I'll land this one and debug the perf/compilation time regress for the PR tuning partitioner.

shunting314 commented:
@pytorchbot merge
